Add mechanism to retrieve API reference page link based on metadata

dipesh-rawat opened the pull request:
Added new **partial layout** named `api-reference-links.html` to fetch the link of the API reference page based on metadata present in the front matter of the file and add the API reference links to page links section of the page. Also, implemented `page-api-reference` shortcode to display relevant API links in the markdown pages.

Fixes #24441

#### Summary of changes

- The `api-reference-links.html` _partial layout_ functionality relies on the presence of `api_metadata` in the front matter of concept pages. This metadata includes "apiVersion" and "kind" attributes, similar to those found in the auto-generated API reference file, Example structure:

```
# Single API
- api_metadata:
  apiVersion: "v1"
  kind: "Node"
```

```
# Mulitple APIs
api_metadata:
- apiVersion: "certificates.k8s.io/v1"
  kind: "CertificateSigningRequest"
- apiVersion: "certificates.k8s.io/v1alpha1"
  kind: "ClusterTrustBundle"  
```

- The partial layout has logic which compares the `apiVersion` and `kind` specified in the current front matter with the metadata in the API reference file located within all the section directories inside [kubernetes-api (here)](https://github.com/kubernetes/website/tree/main/content/en/docs/reference/kubernetes-api/) to identify the correct file for linking.  The existing default layout for docs pages utilizes this partial layout to store all relevant API reference links, which are subsequently used by both `page-meta-link.html` and the `page-api-reference` shortcode.

- Implemented `page-api-reference` shortcode to display relevant API links on the markdown pages, utilizing stored API reference page links from the `api-reference-links.html` partial template. The shortcode includes an optional `kind` parameter to retrieve specific API reference link related to the page.

```
{{< page-api-reference >}}
```

```
{{< page-api-reference kind="ClusterTrustBundle" >}}
```

- **_Extra:_** Additionally, the API reference link is added at the top of the meta links _(along with Edit this page, Create an Issue)_ to facilitate easier access, especially when the Table of Contents is extensive, requiring scrolling to reach the API reference section. _Refer below screenshot_

<details>
<summary>Screenshot (Toggle Me!)</summary>
<img width="1061" alt="api-ref-meta-link" src="https://github.com/kubernetes/website/assets/13051389/b340c6f4-3c82-4757-b327-563776b42f75">
</details>


#### Useful Links

Single API reference link
- [Preview Pod](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/pods/) | [Current Pod](https://kubernetes.io/docs/concepts/workloads/pods/)

Multiple API reference links

- [Preview CSR page](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app//docs/reference/access-authn-authz/certificate-signing-requests/) | [Current CSR page](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/)

Workloads

[Preview Pod](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/pods/) | [Current Pod](https://kubernetes.io/docs/concepts/workloads/pods/)

[Preview Deployment](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/controllers/deployment/) | [Current Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)

[Preview ReplicaSet](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/controllers/replicaset) | [Current ReplicaSet](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/)

Service

[Preview Service](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/services-networking/service/) | [Current Service](https://kubernetes.io/docs/concepts/services-networking/service/)

[Preview Ingress](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/services-networking/ingress/) | [Current Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/)

Storage

[Preview Volume](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/storage/volumes/) | [Current Volume](https://kubernetes.io/docs/concepts/storage/volumes/)

Configruation

[Preview ConfigMap](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/configuration/configmap/) | [Current ConfigMap](https://kubernetes.io/docs/concepts/configuration/configmap/)
[Preview Secret](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/configuration/secret/) | [Current Secret](https://kubernetes.io/docs/concepts/configuration/secret/)

Policy

[Preview LimitRange](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/policy/limit-range/) | [Current LimitRange](https://kubernetes.io/docs/concepts/policy/limit-range/)
[Preview ResourceQuota](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/policy/resource-quotas/) | [Current ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/)

Extend

[Preview CustomResourceDefinitions](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/extend-kubernetes/api-extension/custom-resources/) | [Current CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)


/area web-development

dipesh-rawat commented:
/cc @sftim

I would really appreciate your review and feedback on this whenever you get a chance. Thanks!

sftim commented on data/i18n/en/en.toml:
```diff
@@ -3,6 +3,14 @@
+# NOTE: <resource-api-ref> is a placeholder for the text set by 'api-reference-link' shortcode.
+# Please do not localize or modify <resource-api-ref> placeholder.
+[api_reference_text]
+other = "Please refer to the <resource-api-ref> page for more details."
```
You could use `%s` and a Hugo `printf` instead (I would).

dipesh-rawat replied:
I've updated this in the recent commits as part of the updated approach.

sftim commented on layouts/shortcodes/api-reference-link.html:
```diff
@@ -0,0 +1,41 @@
+  <h2 id="api-reference">{{- $apiReferenceText -}}</h2>
+  <p>
+  {{- $conceptPageApiRefSection := T "api_reference_text" -}}
+  {{- $conceptPageApiRefSection = replace $conceptPageApiRefSection "<resource-api-ref>" $apiReferenceLinks -}}
```
I'd use a printf here; that will fail at render time if there is no placeholder in the string for a localization.

dipesh-rawat replied:
I've revised the text now. Let me know if it needs further enhancement.

sftim commented on static/js/api-reference-link.js:
Can we use static rendering for this? We use Hugo to change the menu already; see `auto_generated_edit_notice` and where we use it.

dipesh-rawat replied:
I've removed this script _(for now)_  as part of the updated approach.

sftim commented on content/en/docs/concepts/architecture/leases.md:
```diff
@@ -1,5 +1,9 @@
 ---
 title: Leases
+api_metadata:
+  apiVersion: "coordination.k8s.io/v1"
+  kind: "Lease"
+  type: cluster
```
I'd make this a _list_ (because a concept might apply to something with two stable API versions, such as  HorizontalPodAutoscaler or Event)

eg (note the extra initial `-`)
```diff
-  apiVersion: "coordination.k8s.io/v1"
+- apiVersion: "coordination.k8s.io/v1"
   kind: "Lease"
-  type: cluster
+  api_reference_section: cluster
```

sftim replied:
If you want a hint about how to work out the `api_reference_section` automatically, let me know. I can help.

dipesh-rawat replied:
I've made changes to implement this in the recent commits.

sftim reviewed (commented):
Thanks for working on this.

- IMO, we can search all the sections of the API reference (so that, as a docs contributor, you don't need to specify them yourself)
  - ie, we'd drop `type: config-and-storage`
- Maybe we don't even need a shortcode? Would a layout `partial` be enough?

sftim commented on content/en/docs/concepts/architecture/nodes.md:
```diff
@@ -570,6 +574,9 @@ see the blog-post about [Kubernetes 1.28: NodeSwap graduates to Beta1](/blog/202
 [KEP-2400](https://github.com/kubernetes/enhancements/issues/4128) and its
 [design proposal](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md).
 
+<!-- Below shortcode creates a "API reference" section linking to the relevant API reference for this page. -->
+{{< api-reference-link >}}
+
 ## {{% heading "whatsnext" %}}
```
I'd put the link inside the further reading, eg:

```suggestion
## {{% heading "whatsnext" %}}

- {{% page-api-reference %}}
```

and possibly with the option to do:
```suggestion
## {{% heading "whatsnext" %}}

Read about the Node API:
- {{% page-api-reference kind="Node" %}}
```

(in case we need to be explicit about what we're linking to)

sftim commented:
@dipesh-rawat I'd be really happy to see progress here; bear in mind that with KubeCon imminent, feedback might take a while.

This improvement is something that could make our docs much more useable for many more people and I look forwards to a tweaked version merging in to _main_.

sftim commented on layouts/shortcodes/api-reference-link.html:
```diff
@@ -0,0 +1,41 @@
+    {{- with .Params.api_metadata -}}
+      {{- if and (eq .kind $kind) (eq .apiVersion $version) -}}
+        <!-- If the file's metadata matches, add link to variable -->
+        {{- $link := printf "<a id='api-reference-page-link' href='%s'>%s %s</a>" $apiRefFile.Permalink $kind $apiReferenceText -}}
```
Maybe use a `class` not `id`. It's possible that one page might link to two APIs; for example, https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/ references CertificateSigningRequest _and_ ClusterTrustBundle.

dipesh-rawat commented:
I'm marking this as WIP now. Once the approach is finalized, I'll incorporate metadata into all the concept pages to leverage this change. I'll also update the PR descriptions accordingly and then remove the WIP label.

/retitle [WIP] Add mechanism to retrieve API reference page link based on metadata

dipesh-rawat commented on content/en/docs/reference/access-authn-authz/certificate-signing-requests.md:
```diff
@@ -5,6 +5,13 @@ reviewers:
+api_metadata:
+- apiVersion: "certificates.k8s.io/v1"
+  kind: "CertificateSigningRequest"
+  override_link_text: "CSR v1"
```
I've included the `override_link_text` feature to customize the default text that appears within the link. This is particularly useful, for instance in the case of HorizontalPodAutoscaler, where we prefer the link to display "v1" and "v2" instead of "HorizontalPodAutoscaler" for both links.

dipesh-rawat commented on layouts/docs/api-reference-links.html:
```diff
@@ -0,0 +1,72 @@
+      <!-- Store the value of $apiReferenceLinks to display in meta link section in 'partials/page-meta-links.html' -->
+      {{ $.Store.Add "apiReferenceLinks" $apiReferenceLinks }}
+      
+      <h2 id="api-reference">{{- $apiReferenceText -}}</h2>
```
A point worth mentioning is that while the new section containing [API reference links ](https://deploy-preview-45496--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/pods/#api-reference) will be added to the end of the page, the section title link is not included in the page's [Table of Contents](https://gohugo.io/methods/page/tableofcontents/). Similar to before, I can incorporate a javascript to append the newly added section to the Table of Contents, or alternatively, we could consider _omitting_ the addition of new section from the page altogether since we now have API reference links in the meta link section above the TOC.  I'm also open to any other suggestions or approaches.

dipesh-rawat commented:
@sftim Apologies for the delay; it took longer than anticipated. Following the last review and feedback, I have updated the implementation to use _layout_ instead of shortcode, addressing the feedback and considering other reviewer comments. Please let me know any further suggestions for improvement.

sftim reviewed (commented):
I like this even more.

I think we can avoid needing a custom layout; instead, change the existing default layout for docs pages to enable the `page-meta-links.html` partial, whenever the `api_metadata` param is present.

We might want to use a mixture of a shortcode and a partial (I probably would):
- always include the link in the right hand navigation
- have a shortcode that links to the same API reference page and automatically looks up the right page to link to
  - make that shortcode optional, and roll it out gradually
- no JavaScript
- no special layout (make it part of the docs default, `single.html`)

sftim commented on layouts/docs/api-reference-links.html:
```diff
@@ -0,0 +1,72 @@
+    {{- end -}}
+
+      <!-- Display API reference heading and links if there are any -->
+      {{- if gt (len $apiReferenceLinks) 0 -}}
```
or, switch to using a shortcode for this optional part

dipesh-rawat commented:
@sftim I've made updates based on the feedback from the last review round. Could you please take another look and provide any further suggestions for improvement? Thank you for your time and inputs!

sftim commented:
/retitle Add mechanism to retrieve API reference page link based on metadata

sftim reviewed (commented):
My only suggestion: split this into 2 commits

1. Add the new mechanism
2. Start to use the new mechanism

That will help localization teams a bit, and also makes a revert of commit 2 easier to do if needed.

dipesh-rawat commented:
> I can see the change, but now the second commit seems to be missing 

I was testing the addition of metadata to some of the concept pages. As requested, I've split this into two commits.

sftim commented:
I'm happy to approve this; it's a change to the live site, so someone else should add LGTM.

sftim reviewed (approved):
/approve

sftim commented on content/en/docs/concepts/services-networking/ingress.md:
aside: we could also link to IngressClass

divya-mohan0209 reviewed (commented):
Thanks @dipesh-rawat !

This LGTM :)
/lgtm

sftim commented:
Nice. I'll send in a PR to make a similar change, to try it out.
